File: backend/spec/factories/share_holdings.rb
# frozen_string_literal: true

FactoryBot.define do
  factory :share_holding do
    company_investor
    company_investor_entity do
      association :company_investor_entity,
                  company: company_investor.company,
                  name: company_investor.user.legal_name
    end
    share_class { create(:share_class, company: company_investor.company) }
    sequence(:name) { |n| "GUM-#{n}" }
    issued_at { 1.year.ago }
    originally_acquired_at { issued_at }
    number_of_shares { 101 }
    share_price_usd { 111.22 }
    total_amount_in_cents { (number_of_shares * (share_price_usd * 100.0)).round }
    share_holder_name { company_investor.user.legal_name }
  end
end

File: e2e/factories/shareHoldings.ts
import { companyInvestors, shareClasses, shareHoldings } from "@/db/schema";
import { assert } from "@/utils/assert";
import { db, takeOrThrow } from "@test/db";
import { companyInvestorsFactory } from "@test/factories/companyInvestors";
import { shareClassesFactory } from "@test/factories/shareClasses";
import { eq } from "drizzle-orm";

export const shareHoldingsFactory = {
  create: async (overrides: Partial<typeof shareHoldings.$inferInsert> = {}) => {
    const investor = overrides.companyInvestorId
      ? await db.query.companyInvestors
          .findFirst({ where: eq(companyInvestors.id, overrides.companyInvestorId) })
          .then(takeOrThrow)
      : (await companyInvestorsFactory.create(overrides)).companyInvestor;
    const shareClass = overrides.shareClassId
      ? await db.query.shareClasses.findFirst({ where: eq(shareClasses.id, overrides.shareClassId) }).then(takeOrThrow)
      : (await shareClassesFactory.create({ companyId: investor.companyId })).shareClass;
    const [shareHolding] = await db
      .insert(shareHoldings)
      .values({
        companyInvestorId: investor.id,
        shareClassId: shareClass.id,
        name: `C2-1`,
        numberOfShares: 100,
        issuedAt: new Date(),
        originallyAcquiredAt: new Date(),
        totalAmountInCents: BigInt(10000),
        sharePriceUsd: "10.00",
        shareHolderName: "Test Holder",
        ...overrides,
      })
      .returning();
    assert(shareHolding != null);
    return shareHolding;
  },
};
